common/src/block_interval.rs

```diff
@@ -0,0 +1,255 @@
+use alloy::primitives::B256;
+use alloy::rpc::types::eth::BlockId;
+use alloy::{hex, providers::Provider, transports::Transport};
+use async_stream::try_stream;
+use thiserror::Error;
+use tokio_stream::Stream;
+use tracing::{error, info};
+
+use crate::parsing;
+
+#[derive(Error, Debug, PartialEq)]
+pub enum BlockIntervalError {
+    #[error("invalid block number '{0}'")]
+    InvalidBlockNumber(String),
+    #[error("invalid follow-from range '{0}', should be `lhs..`")]
+    InvalidFollowFromRange(String),
+    #[error("invalid block interval range '{0}'")]
+    InvalidBlockIntervalRange(String),
+    #[error("could not create bounded stream from unbounded follow-from interval")]
+    IntoBoundedStreamError,
+    #[error("could not create unbounded follow-from stream from fixed bounded interval")]
+    IntoUnboundedStreamError,
+    #[error("could not retrieve latest block number from the provider: {0}")]
+    LatestBlockNumberError(String),
+    #[error("invalid block hash '{0}'")]
+    InvalidBlockHash(String),
+}
+
+/// Range of blocks to be processed and proven.
+#[allow(unused)]
+#[derive(Debug, PartialEq, Clone)]
+pub enum BlockInterval {
+    // A single block number.
+    Single(u64),
+    // A single block id (could be number or hash)
+    SingleBlockId(BlockId),
+    // A range of blocks.
+    Range(std::ops::Range<u64>),
+    // Starting block for the continuous stream of blocks,
+    // acquired from the blockchain node.
+    FollowFrom(u64),
+}
+
+impl BlockInterval {
+    /// Create a new block interval
+    ///
+    /// A valid block range is of the form:
+    ///     * `block_number` for a single block number
+    ///     * `lhs..rhs`, `lhs..=rhs` as an exclusive/inclusive range
+    ///     * `lhs..` for a range starting from `lhs` to the chain tip. `lhs..=`
+    ///       is also valid format.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    ///    # use common::block_interval::BlockInterval;
+    ///    assert_eq!(BlockInterval::new("0..10").unwrap(), BlockInterval::Range(0..10));
+    ///    assert_eq!(BlockInterval::new("0..=10").unwrap(), BlockInterval::Range(0..11));
+    ///    assert_eq!(BlockInterval::new("32141").unwrap(), BlockInterval::Single(32141));
+    ///    assert_eq!(BlockInterval::new("100..").unwrap(), BlockInterval::FollowFrom(100));
+    /// ```
+    pub fn new(s: &str) -> Result<BlockInterval, BlockIntervalError> {
```

Comment:
See comment on `parsing`.

Also this should be a [`FromStr`](https://doc.rust-lang.org/std/str/trait.FromStr.html) implementation, with a simple `anyhow::Error`, and should delegate to `BlockId::from_str` where appropriate

Reply:
I will add additional `FromStr` implementation.